File: stacks/go/base/code/observability.tpl.go
// File: observability.tpl.go
// Purpose: OpenTelemetry tracing setup for Go
// Generated for: {{PROJECT_NAME}}

package observability

import (
	"context"
	"log"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
	"go.opentelemetry.io/otel/trace"
)

func SetupTracing(ctx context.Context, serviceName, otlpEndpoint string) (func(context.Context) error, error) {
	exporter, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpoint(otlpEndpoint),
		otlptracegrpc.WithInsecure(),
	)
	if err != nil {
		return nil, err
	}

	res, err := resource.Merge(
		resource.Default(),
		resource.NewWithAttributes(
			semconv.SchemaURL,
			semconv.ServiceName(serviceName),
		),
	)
	if err != nil {
		return nil, err
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
	)

	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	return tp.Shutdown, nil
}

func GetTracer(name string) trace.Tracer {
	return otel.Tracer(name)
}

// WithSpan is a helper for manual tracing
func WithSpan(ctx context.Context, name string, fn func(ctx context.Context) error, attrs ...attribute.KeyValue) error {
	tracer := GetTracer("default")
	ctx, span := tracer.Start(ctx, name)
	defer span.End()

	span.SetAttributes(attrs...)

	if err := fn(ctx); err != nil {
		span.RecordError(err)
		return err
	}

	return nil
}

// Example usage
func ExampleUsage() {
	ctx := context.Background()
	shutdown, err := SetupTracing(ctx, "my-service", "localhost:4317")
	if err != nil {
		log.Fatal(err)
	}
	defer shutdown(ctx)

	tracer := GetTracer("my-service")
	ctx, span := tracer.Start(ctx, "operation")
	defer span.End()

	span.SetAttributes(attribute.String("user.id", "123"))
}

File: stacks/python/base/code/observability.tpl.py
"""
File: observability.tpl.py
Purpose: OpenTelemetry tracing setup
Generated for: {{PROJECT_NAME}}
"""

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource, SERVICE_NAME
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor
from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor

def setup_telemetry(service_name: str, otlp_endpoint: str = "http://localhost:4317"):
    """Initialize OpenTelemetry tracing"""
    
    resource = Resource.create({SERVICE_NAME: service_name})
    
    provider = TracerProvider(resource=resource)
    
    otlp_exporter = OTLPSpanExporter(endpoint=otlp_endpoint, insecure=True)
    processor = BatchSpanProcessor(otlp_exporter)
    provider.add_span_processor(processor)
    
    trace.set_tracer_provider(provider)
    
    # Auto-instrument common libraries
    RequestsInstrumentor().instrument()
    
    return trace.get_tracer(service_name)

def instrument_fastapi(app):
    """Instrument FastAPI application"""
    FastAPIInstrumentor.instrument_app(app)

def instrument_sqlalchemy(engine):
    """Instrument SQLAlchemy engine"""
    SQLAlchemyInstrumentor().instrument(engine=engine)

# Manual tracing example
def example_tracing():
    tracer = trace.get_tracer(__name__)
    
    with tracer.start_as_current_span("parent-operation") as span:
        span.set_attribute("user.id", "123")
        
        with tracer.start_as_current_span("child-operation"):
            # Do work here
            pass

# Usage:
# tracer = setup_telemetry("my-service")
# instrument_fastapi(app)
